presto-main/src/main/java/com/facebook/presto/operator/aggregation/AbstractMinMaxAggregation.java

```diff
@@ -0,0 +1,179 @@
+    private static final MethodHandle BOOLEAN_INPUT_FUNCTION = methodHandle(BooleanMinMaxAggregationFunctions.class, "input", MethodHandle.class, NullableBooleanState.class, boolean.class);
+    private static final MethodHandle BOOLEAN_COMBINE_FUNCTION = methodHandle(BooleanMinMaxAggregationFunctions.class, "combine", MethodHandle.class, NullableBooleanState.class, NullableBooleanState.class);
+    private static final MethodHandle BOOLEAN_OUTPUT_FUNCTION = methodHandle(BooleanMinMaxAggregationFunctions.class, "output", Type.class, NullableBooleanState.class, BlockBuilder.class);
+
+    private final String name;
+    private final OperatorType operatorType;
+    private final Signature signature;
+
+    private final StateCompiler compiler = new StateCompiler();
+
+    protected AbstractMinMaxAggregation(String name, OperatorType operatorType)
+    {
+        checkNotNull(name);
+        checkNotNull(operatorType);
+        this.name = name;
+        this.operatorType = operatorType;
+        this.signature = new Signature(name, ImmutableList.of(orderableTypeParameter("E")), "E", ImmutableList.of("E"), false, false);
+    }
+
+    @Override
+    public Signature getSignature()
+    {
+        return signature;
+    }
+
+    @Override
+    public FunctionInfo specialize(Map<String, Type> types, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
+    {
+        Type type = types.get("E");
+        MethodHandle greaterThanMethodHandle = functionRegistry.resolveOperator(operatorType, ImmutableList.of(type, type)).getMethodHandle();
+        Signature signature = new Signature(name, type.getTypeSignature(), type.getTypeSignature());
+        InternalAggregationFunction aggregation = generateAggregation(type, greaterThanMethodHandle);
+        return new FunctionInfo(signature, getDescription(), aggregation.getIntermediateType().getTypeSignature(), aggregation, false);
+    }
+
+    protected InternalAggregationFunction generateAggregation(Type type, MethodHandle compareMethodHandle)
+    {
+        DynamicClassLoader classLoader = new DynamicClassLoader(AbstractMinMaxAggregation.class.getClassLoader());
+
+        List<Type> inputTypes = ImmutableList.of(type);
+
+        AccumulatorStateSerializer stateSerializer;
+        AccumulatorStateFactory stateFactory;
+        MethodHandle inputFunction;
+        MethodHandle combineFunction;
+        MethodHandle outputFunction;
+        Class<? extends AccumulatorState> stateInterface;
+
+        if (type.getJavaType() == long.class) {
+            stateFactory = compiler.generateStateFactory(NullableLongState.class, classLoader);
+            stateSerializer = new NullableLongStateSerializer();
+            stateInterface = NullableLongState.class;
+            inputFunction = LONG_INPUT_FUNCTION;
+            combineFunction = LONG_COMBINE_FUNCTION;
```

Comment:
There's an optimization we can make here. The aggregation framework supports "intermediate input functions". These can be used instead of the combine function when the input and the intermediate input (the result of serializing a state object) are the same type.

So what you can do is use the input function as both the input, and the intermediate input function, and then remove the combine function.

You'll also need to add `type` as a parameter to the serializers, use that when serializing, and return it as the "intermediate type"